src/backend/utils/cache/plancache.c

```diff
@@ -1073,12 +1085,16 @@ choose_custom_plan(CachedPlanSource *plansource, ParamListInfo boundParams, Into
 /*
  * cached_plan_cost: calculate estimated cost of a plan
  *
- * If include_planner is true, also include the estimated cost of constructing
+ * When generating custom plan, we need to know, will it be safe to compare
+ * its cost with generic plan cost on the second choose_custom_plan() call.
+ * If is_custom_cost_safe ptr is not NULL, returning false will indicate query
+ * planned using any optimizer except Postgres optimizer. If function called
+ * with is_custom_cost_safe, we also include the estimated cost of constructing
  * the plan.  (We must factor that into the cost of using a custom plan, but
  * we don't count it for a generic plan.)
  */
 static double
-cached_plan_cost(CachedPlan *plan, bool include_planner)
+cached_plan_cost(CachedPlan *plan, bool *is_custom_cost_safe)
```

Comment:
I don't like an idea of adding an additional loop just to calculate `is_custom_cost_safe` flag.
As for semantic meaning, why can't we change it here a bit? Calculating this flag is very near the all work done in this and adjacent functions.

Reply:
> As for semantic meaning, why can't we change it here a bit?

The direct semantics of `cached_plan_cost()` is about computing cost and your additional parameter injection alludes to having some side effect that checks whether ORCA optimizer was used. This check is activated not on each call of `cached_plan_cost()` but when we pass false value to the old `include_planner` parameter despite of this parameter and `is_custom_cost_safe` definition logic are not related semantically. Moreover, this is wrong behavior because when we first time build generic plan in the following branch:
```C
CachedPlan *
GetCachedPlan(CachedPlanSource *plansource, ParamListInfo boundParams,
			  bool useResOwner, IntoClause *intoClause)
{
    ...

    /* Build a new generic plan */
    plan = BuildCachedPlan(plansource, qlist, NULL, NULL);
    ...
    /* Update generic_cost whenever we make a new generic plan */
    plansource->generic_cost = cached_plan_cost(plan, NULL);

    /*
     * If, based on the now-known value of generic_cost, we'd not have
     * chosen to use a generic plan, then forget it and make a custom
     * plan.  This is a bit of a wart but is necessary to avoid a
     * glitch in behavior when the custom plans are consistently big
     * winners; at some point we'll experiment with a generic plan and
     * find it's a loser, but we don't want to actually execute that
     * plan.
     */
    customplan = choose_custom_plan(plansource, boundParams, intoClause);

    ...
}
```
we will use old (on previous iteration defined) value of `is_custom_cost_safe` flag in `choose_custom_plan()` call.

Reply:
> We don't replace Param node with Const Node if boundParams are NULL and fall at Param node processing as shown above.

Do you want to point out that generic plan built after call of `BuildCachedPlan()` with NULL `boundParams` argument always contains `Param` nodes (I suppose, denoting external parameters of prepared query) and we result in fallback to postgres optimizer?

Reply:
> As for extending semantics of argument and function, in voice chat we decided to wait for third side review.

Not only this. The third side review have to define whether we have to move the calculation of `is_custom_cost_safe` logic upper into `GetCachedPlan()` function?

Reply:
> Not only this. The third side review have to define whether we have to move the calculation of is_custom_cost_safe logic upper into GetCachedPlan() function?
All this staff.

Reply:
> Do you want to point out that generic plan built after call of BuildCachedPlan() with NULL boundParams argument always contains Param nodes (I suppose, denoting external parameters of prepared query) and we result in fallback to postgres optimizer?

No. There can be no `Param` nodes, so calling `BuildCachedPlan()` with NULL `boundParams` can return statements generated with `PLANGEN_OPTIMIZER`. Costs comparison should be reworked.

Reply:
I vote for @maksm90 solution with an iteration over `plan->stmt_list` in `GetCachedPlan()`. The code is not only about performance, but also about maintainability. The current solution adds complicity to the code, the original approach with `include_planner` is more obvious. Besides, I vote to preserve PostgreSQL API as much as possible (master branch still uses `cached_plan_cost(CachedPlan *plan, bool include_planner)`.